clash-prelude/src/Clash/Explicit/BlockRam.hs

```diff
@@ -1038,3 +1046,241 @@ readNew clk rst en ram rdAddr wrM = mux wasSame wasWritten $ ram rdAddr wrM
         (wasSame,wasWritten) =
           unbundle (register clk rst en (False, undefined)
                              (readNewT <$> rdAddr <*> wrM))
+
+-- | Produces vendor-agnostic HDL that will be inferred as a true, dual port
+-- block ram. It has write-first (or "new data") behavior, no clock enables,
+-- no byte enables, and is symmetric in its data ports.
+--
+-- * __NB__: Read value is delayed by 1 cycle
+-- * __NB__: Initial output value is /undefined/
+-- * __NB__: Port A and port B cannot make overlapping read/write requests on
+--           the same address.
+--
+trueDualPortBlockRam ::
+  forall nAddrs domA domB a .
+  ( HasCallStack
+  , KnownNat nAddrs
+  , KnownDomain domA
+  , KnownDomain domB
+  , KnownNat (DomainPeriod domA)
+  , KnownNat (DomainPeriod domB)
+  , NFDataX a
+  , BitPack a
+  ) =>
+
+  -- | Clock for port A
+  Clock domA ->
+  -- | Write data port A
+  Signal domA (Maybe a) ->
+  -- | Address to read from or write to on port A
+  Signal domA (Index nAddrs) ->
+
+  -- | Clock for port B
+  Clock domB ->
+  -- | Write data B
+  Signal domB (Maybe a) ->
+  -- | Address to read from or write to on port B
+  Signal domB (Index nAddrs) ->
+
+  -- | Outputs data on /next/ cycle. If write enable is @True@, the data written
+  -- will be echoed. If write enable is @False@, the read data is returned.
+  (Signal domA a, Signal domB a)
+trueDualPortBlockRam clkA writeA addrA clkB writeB addrB = case compareSNat (SNat @(DomainPeriod domB)) (SNat @(DomainPeriod domA)) of
+  SNatLE ->
+    trueDualPortBlockRam#
+      clkA (isJust <$> writeA) addrA (fromJustX <$> writeA)
+      clkB (isJust <$> writeB) addrB (fromJustX <$> writeB)
+  SNatGT ->
+    swap (trueDualPortBlockRam clkB writeB addrB clkA writeA addrA)
+
+toMaybeX :: a -> MaybeX a
+toMaybeX a =
+  case isX a of
+    Left _ -> IsX
+    Right _ -> IsDefined a
+
+data MaybeX a = IsX | IsDefined !a
```

Comment:
I added this type, but wasn't sure whether it was really necessary. Perhaps a `Maybe` + documentation in the record would do. I like the explicitness of it though.